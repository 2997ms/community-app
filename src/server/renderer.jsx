/**
 * This module implements ExpressJS middleware for server-side rendering of
 * the App.
 */

import config from 'config';
import React from 'react';
import ReactDOM from 'react-dom/server';
import { StaticRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import serializeJs from 'serialize-javascript';
import { isDev } from 'utils/isomorphy';

import App from '../shared';

/* This is always initial state of the store. Later we'll have to provide a way
 * to put the store into correct state depending on the demanded route. */
import storeFactory from '../shared/store-factory';

export default (req, res) => {
  storeFactory(req).then((store) => {
    const context = {};
    const appHtml = ReactDOM.renderToString((
      <Provider store={store}>
        <StaticRouter
          context={context}
          location={req.url}
        >
          <App />
        </StaticRouter>
      </Provider>
    ));
    if (context.status) res.status(context.status);
    res.send((
      `<!DOCTYPE html>
      <html>
        <head>
          <title>Topcoder</title>
          ${
            /* In dev environment styles are embed into JS bundle,
              and we should not include this link into the page, otherwise
              the page will initially load production /style.css and it may
              result in some wierd behavior while loading. */
            isDev() ? '' : '<link rel="stylesheet" href="/style.css" />'
          }
        </head>
        <body>
          <div id="react-view">${appHtml}</div>
          <script type="application/javascript">
            window.CONFIG = ${serializeJs(config, { isJSON: true })}
            window.ISTATE = ${serializeJs(store.getState(), { isJSON: true })}
          </script>
          <script type="application/javascript" src="/bundle.js"></script>
        </body>
      </html>`
    ));
  });
};
