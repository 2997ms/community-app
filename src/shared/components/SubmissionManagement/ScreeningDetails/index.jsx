/**
 * This compnent receives via props the screening object and nicely renders them.
 * Note that this component both has the text coming directly from the screening object,
 * and the text generated based on the screening status
 *
 */

import _ from 'lodash';
import Button from 'components/Button';
import React, { PropTypes as PT } from 'react';
import shortid from 'shortid';
import './styles.scss';

export default function ScreeningDetails(props) {
  const {
    screeningObject,
    onHelp,
    onOpenOnlineReview,
    submissionId,
  } = props;

  const checkScreeningObject = () => {
    if (screeningObject) {
      return screeningObject;
    }
    return {};
  };
  const hasWarnings = checkScreeningObject(screeningObject).warnings;
  const hasStatus = checkScreeningObject(screeningObject).status;
  const hasStatusPassed = hasStatus === 'passed';
  const hasStatusFailed = hasStatus === 'failed';
  const hasPending = checkScreeningObject(screeningObject).status === 'pending';
  const warnLength = checkScreeningObject(screeningObject).warnings && hasWarnings.length;

  const setStatusInfo = () => {
    if (hasPending) {
      return {
        title: 'Pending',
        classname: 'pending',
        message: 'Your submission has been received, and will be screened after the end of the phase',
      };
    } else if (hasStatusPassed && !hasWarnings) {
      return {
        title: 'Passed Screening',
        classname: 'passed',
        message: 'You have passed screening.',
      };
    } else if (hasStatusFailed && !hasWarnings) {
      return {
        title: 'Failed Screening',
        classname: 'failed',
        message: 'You have failed screening',
      };
    } else if (hasStatusPassed && hasWarnings) {
      return {
        title: 'Passed Screening with Warnings',
        classname: 'passed',
        message: `You have passed screening, but the screener has given you ${warnLength} warnings that you must fix in round 2.`,
      };
    } else if (hasStatusFailed && hasWarnings) {
      return {
        title: 'Failed Screening with Warnings',
        classname: 'failed',
        message: 'You have failed screening and the screener has given you the following warning.',
      };
    }
    return {
      title: '',
      classname: '',
      message: 'Your submission has been received, and will be evaluated during Review phase.',
    };
  };

  let warnings = [];
  if (checkScreeningObject(screeningObject).warnings) {
    warnings = screeningObject.warnings.map((warning, i) =>
      <div styleName="screening-warning" key={shortid.generate()}>
        <div styleName="warning-bold"><span>Warning</span> {`${1 + i} : ${warning.brief}`}</div>
        <p>{warning.details}</p>
      </div>,
    );
  }
  return (
    <div styleName="screening-details" >
      <div styleName="screening-details-head">
        <p styleName={`status-title ${setStatusInfo().classname}`}>{setStatusInfo().title}</p>
        <button
          styleName="online-review-link"
          onClick={() => onOpenOnlineReview(submissionId)}
        >Online Review</button>
      </div>
      <p>{setStatusInfo().message}</p>
      <div styleName="screening-warning">
        {warnings}
        {((hasStatusFailed) || (hasStatusPassed && hasWarnings)) &&
          <p styleName="more-info">Need more info on how to pass screening?
             Go to help to read Rules & Policies.</p>}
        <div styleName="help-btn">
          <Button className="tc-btn-default" onClick={() => onHelp(submissionId)}>Help</Button>
        </div>
      </div>
    </div>
  );
}

ScreeningDetails.defaultProps = {
  screeningObject: {},
  onHelp: _.noop,
  onOpenOnlineReview: _.noop,
};

ScreeningDetails.propTypes = {
  screeningObject: PT.shape({
    status: PT.string,
    warnings: PT.arrayOf(PT.shape({
      brief: PT.string.isRequired,
      details: PT.string.isRequired,
    })),
  }),
  onHelp: PT.func,
  onOpenOnlineReview: PT.func,
  submissionId: PT.string.isRequired,
};
