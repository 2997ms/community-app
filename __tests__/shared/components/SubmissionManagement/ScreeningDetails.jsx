import React from 'react';
import renderer from 'react-test-renderer';

import ScreeningDetails from 'components/SubmissionManagement/ScreeningDetails';
import mockScreeningObject from './__mocks__/screening-object.json';

test('renders correctly', () => {
  const html = renderer.create((
    <ScreeningDetails
      screeningObject={mockScreeningObject}
      submissionId="12345"
    />
  )).toJSON();
  expect(html).toMatchSnapshot();
});
