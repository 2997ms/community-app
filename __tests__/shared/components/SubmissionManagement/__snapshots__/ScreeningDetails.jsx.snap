// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="src-shared-components-SubmissionManagement-ScreeningDetails-___styles__screening-details___2ycEB"
>
  <div
    className="src-shared-components-SubmissionManagement-ScreeningDetails-___styles__screening-details-head___2rqCn"
  >
    <p
      className="src-shared-components-SubmissionManagement-ScreeningDetails-___styles__status-title___3HTbG"
    >
      
    </p>
    <button
      className="src-shared-components-SubmissionManagement-ScreeningDetails-___styles__online-review-link___14IJN"
      onClick={[Function]}
    >
      Online Review
    </button>
  </div>
  <p>
    Your submission has been received, and will be evaluated during Review phase.
  </p>
  <div
    className="src-shared-components-SubmissionManagement-ScreeningDetails-___styles__screening-warning___1IMq5"
  >
    <div
      className="src-shared-components-SubmissionManagement-ScreeningDetails-___styles__screening-warning___1IMq5"
    >
      <div
        className="src-shared-components-SubmissionManagement-ScreeningDetails-___styles__warning-bold___20osd"
      >
        <span>
          Warning
        </span>
         
        1 : Dummy Brief #1
      </div>
      <p>
        Dummy Details #1
      </p>
    </div>
    <div
      className="src-shared-components-SubmissionManagement-ScreeningDetails-___styles__screening-warning___1IMq5"
    >
      <div
        className="src-shared-components-SubmissionManagement-ScreeningDetails-___styles__warning-bold___20osd"
      >
        <span>
          Warning
        </span>
         
        2 : Dummy Brief #2
      </div>
      <p>
        Dummy Details #2
      </p>
    </div>
    <div
      className="src-shared-components-SubmissionManagement-ScreeningDetails-___styles__help-btn___222Uf"
    >
      <button
        className="tc-btn tc-btn-default"
        onClick={[Function]}
      >
        Help
      </button>
    </div>
  </div>
</div>
`;
